File: aip/packages/python/src/aip/sdk/http.py
"""HMAC-SHA256 signing and verification for AIP webhook delivery.

Port of packages/typescript/src/sdk/http.ts
"""

from __future__ import annotations

import hashlib
import hmac


def sign_payload(secret: str, payload: str) -> str:
    """Sign a payload with HMAC-SHA256. Returns ``sha256={hex}``."""
    digest = hmac.new(
        secret.encode("utf-8"),
        payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    return f"sha256={digest}"


def verify_signature(secret: str, payload: str, signature: str) -> bool:
    """Verify an HMAC-SHA256 signature using constant-time comparison.

    SPEC requires constant-time comparison to prevent timing attacks.
    """
    expected = sign_payload(secret, payload)
    return hmac.compare_digest(expected, signature)

File: aip/packages/typescript/src/sdk/http.ts
/**
 * HMAC-SHA256 signing and verification for AIP webhook delivery.
 *
 * Uses Node.js crypto (available in Node 18+ and Cloudflare Workers).
 */

import { createHmac } from "node:crypto";

/** Sign a payload with HMAC-SHA256. Returns `sha256={hex}`. */
export function signPayload(secret: string, payload: string): string {
  const hmac = createHmac("sha256", secret);
  hmac.update(payload);
  return `sha256=${hmac.digest("hex")}`;
}

/**
 * Verify an HMAC-SHA256 signature using constant-time comparison.
 * SPEC requires constant-time comparison to prevent timing attacks.
 */
export function verifySignature(
  secret: string,
  payload: string,
  signature: string
): boolean {
  const expected = signPayload(secret, payload);
  return constantTimeEqual(expected, signature);
}

function constantTimeEqual(a: string, b: string): boolean {
  const maxLen = Math.max(a.length, b.length);
  const aPadded = a.padEnd(maxLen, '\0');
  const bPadded = b.padEnd(maxLen, '\0');
  let result = 0;
  for (let i = 0; i < maxLen; i++) {
    result |= aPadded.charCodeAt(i) ^ bPadded.charCodeAt(i);
  }
  // Also check lengths match (after constant-time comparison)
  result |= a.length ^ b.length;
  return result === 0;
}
